url47fcfc0133_baliosian_tau_fst_tgz-pJ8-uy_edu_fing_mina_fsa_test_old_TestComposition1J8/src/uy/edu/fing/mina/lupa/LupaExporterRatePower.java

```diff
@@ -0,0 +1,1027 @@
+					  out = out.concat("\t\tif ret > retMax then\n" +
+				      		"\t\t\tretMax = ret\n" +
+				      		"\t\t\tl = lp\n" +
+				      		"\t\tend\n" +
+				      		"\tend\n" +
+				      		"\treturn notifs.change_power(l,e)\n" +
+					      	"end\n\n");					  
+				  }
+                out = out.concat("\t-- Membership function for " + functionName + "\n");
+				out = out.concat("functions." + functionName + " = function(l)\n");
+				out = out.concat("\t------------------------------------------------\n");
+				out = out.concat("\t-- TODO: Complete this with your function code. --\n");
+				out = out.concat("\t------------------------------------------------\n");
+				out = out.concat("end\n\n");
+				
+
+//                out = out.concat("\t-- Notification generator for " + functionName + "\n");
+//				out = out.concat("notifs." + functionName + " = function(l,e)\n");
+//				out = out.concat("\t------------------------------------------------\n");
+//				out = out.concat("\t-- TODO: Complete this with your notification code. --\n");
+//				out = out.concat("\t------------------------------------------------\n");
+//				out = out.concat("end\n\n");
+			}
+			
+		}
+		return out;
+	}
+
+	/**
+	 * Creates a String with event functions that are CompositeTf or Nots.
+	 * @param tffst The TFFST to export.
+	 * @return The updated String objects with the new function bodies that it may have.
+	 * @throws UnsupportedTFFSTException 
+	 */
+	private static String compositeEventsToLupa(Tffst tffst) throws UnsupportedTFFSTException {
+		String out = new String();
+		
+		out = out.concat("-----------------------------------------------\n");
+		out = out.concat("--          BEGIN COMPOSITE EVENTS           --\n");
+		out = out.concat("-----------------------------------------------\n");
+		CompositeTf comp;
+		TfI leftTf;
+		TfI rightTf;
+		String functionName;
+		for(TfI function : compositeEvents.values()){
+			functionName = functionName(function);
+
+		    if (function.isNot()) {
+				out = out.concat("-- " + function.getName() + "\n");
+				out = out.concat("events." + functionName + " = function(e) \n");
+				TfI nonneg = function.not();
+				out = out.concat("\tlocal nonneg = events." + functionName(nonneg) + "(e)\n");
+				out = out.concat("\treturn 1-nonneg\n");
+				out = out.concat("end\n");
+			}
+		    else if(function instanceof CompositeTf){
+				comp = (CompositeTf) function;
+				leftTf = comp.left;
+				rightTf = comp.right;
+				out = out.concat("-- " + function.getName() + "\n");
+				out = out.concat("events." + functionName + " = function(e) \n");
+//				out = out.concat(functionHeader(functionName) + "\n");
+				out = out.concat("\tlocal left= events." + functionName(leftTf) + "(e)\n");
+				out = out.concat("\tlocal right= events."	+ functionName(rightTf) + "(e)\n");
+				if (comp.op == Operator.AND) {
+					out = out.concat("\tif left < right then return left else return right end\n");
+				} else if (comp.op == Operator.OR) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.